### PR TITLE
keep userData and new grab properties in sync

### PR DESCRIPTION
**Pull request by sethalves**
- keep userData and new grab properties in sync

https://highfidelity.manuscript.com/f/cases/19448

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14219/HighFidelity-Beta-PR14246-c56e36b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14219/HighFidelity-Beta-Interface-PR14246-c56e36b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14219/HighFidelity-Beta-PR14246-c56e36b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14219/HighFidelity-Beta-Interface-PR14246-c56e36b.exe

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 1]
```diff
@@ -236,6 +239,94 @@ EntityItemProperties convertPropertiesFromScriptSemantics(const EntityItemProper
     return entitySideProperties;
 }
 
+void synchronizeEditedGrabProperties(EntityItemProperties& properties) {
+    // After sufficient warning to content creators, we should be able to remove this.
+
+    if (properties.grabbingRelatedPropertyChanged()) {
+        // This edit touches a new-style grab property, so make userData agree...
+        GrabPropertyGroup& grabProperties = properties.getGrab();
+        QJsonObject userData = QJsonDocument::fromJson(properties.getUserData().toUtf8()).object();
```
Comment (huffman) [inline libraries/entities/src/EntityScriptingInterface.cpp]: We should validate that this doesn't fail, especially if userData is not empty. It's not required that the field contains json, in which case we might clobber non-json data.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14227/HighFidelity-Beta-PR14246-5d9870f.apk

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14227/HighFidelity-Beta-Interface-PR14246-5d9870f.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14270/HighFidelity-Beta-PR14246-befdc0e.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14270/HighFidelity-Beta-PR14246-befdc0e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14270/HighFidelity-Beta-Interface-PR14246-befdc0e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14270/HighFidelity-Beta-PR14246-befdc0e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14270/HighFidelity-Beta-Interface-PR14246-befdc0e.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14343/HighFidelity-Beta-PR14246-2989657.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14343/HighFidelity-Beta-PR14246-2989657.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14343/HighFidelity-Beta-Interface-PR14246-2989657.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14343/HighFidelity-Beta-PR14246-2989657.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14343/HighFidelity-Beta-Interface-PR14246-2989657.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14356/HighFidelity-Beta-PR14246-1b557c2.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14356/HighFidelity-Beta-PR14246-1b557c2.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14356/HighFidelity-Beta-Interface-PR14246-1b557c2.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14356/HighFidelity-Beta-PR14246-1b557c2.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14356/HighFidelity-Beta-Interface-PR14246-1b557c2.exe

**Comment by sethalves** [comment]
to test:
- run the tests in https://highfidelity.testrail.net/index.php?/suites/view/1134&group_by=cases:section_id&group_order=asc
- run https://highfidelity.testrail.net/index.php?/cases/view/17764
- generally exercise near and far grabbing and equipping on a variety of content
- open the HMD-friendly script output window and run the attached script.  look for all "ok"s.

[test-grab-prop-sync.js.txt](https://github.com/highfidelity/hifi/files/2512002/test-grab-prop-sync.js.txt)

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14363/HighFidelity-Beta-PR14246-e07fd9f.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14363/HighFidelity-Beta-PR14246-e07fd9f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14363/HighFidelity-Beta-Interface-PR14246-e07fd9f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14363/HighFidelity-Beta-PR14246-e07fd9f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14363/HighFidelity-Beta-Interface-PR14246-e07fd9f.exe

**Comment by JeffClinton** [comment]
Currently testing

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14394/HighFidelity-Beta-PR14246-b8f1c96.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14394/HighFidelity-Beta-PR14246-b8f1c96.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14394/HighFidelity-Beta-Interface-PR14246-b8f1c96.exe

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by qlimnavarro** [comment]
Pass @conklin94122 
All grab tests function as expected.

**Pull request merged**